src/operator/controllers/intents/reconciler.go

```diff
@@ -0,0 +1,322 @@
+		context.Background(),
+		&otterizev1alpha3.ClientIntents{}, OtterizeIntentsClientIndexNameField,
+		func(object client.Object) []string {
+			clientIntents := object.(*otterizev1alpha3.ClientIntents)
+			return []string{fmt.Sprintf("%s.%s", clientIntents.Spec.Service.Name, clientIntents.Namespace)}
+		})
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) InitIntentsDatabaseServerIndices(mgr ctrl.Manager) error {
+	err := mgr.GetCache().IndexField(
+		context.Background(),
+		&otterizev1alpha3.ClientIntents{},
+		otterizev1alpha3.OtterizeTargetServerIndexField,
+		func(object client.Object) []string {
+			var res []string
+			intents := object.(*otterizev1alpha3.ClientIntents)
+			if intents.Spec == nil {
+				return nil
+			}
+			// Only relevant for database intents for now
+			for _, intent := range intents.GetCallsList() {
+				if intent.DatabaseResources != nil {
+					res = append(res, intent.GetTargetServerName())
+				}
+			}
+			return res
+		})
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	return nil
+}
+
+func (r *Reconciler) handleDBUserCreation(
+	ctx context.Context,
+	intents otterizev1alpha3.ClientIntents,
+	pgConfigurator *postgres.PostgresConfigurator,
+	databaseName string) error {
+
+	pgUsername, err := r.getPostgresUserForWorkload(ctx, intents.GetServiceName(), intents.Namespace)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	exists, err := pgConfigurator.ValidateUserExists(ctx, pgUsername)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	if !exists {
+		password, err := r.fetchWorkloadPassword(ctx, intents)
+		if err != nil {
+			r.recorder.Eventf(&intents, v1.EventTypeWarning,
+				ReasonFailedReadingWorkloadPassword,
+				"Failed reading client %s Postgres password. Error: %s", intents.GetServiceName(), err.Error())
+			return errors.Wrap(err)
+		}
+
+		logrus.WithField("username", pgUsername).Infof(
+			"Username does not exist in database %s, creating it", databaseName)
+
+		err = r.createPostgresUserForWorkload(ctx, pgConfigurator, pgUsername, password)
+		if err != nil {
+			r.recorder.Eventf(&intents, v1.EventTypeWarning, ReasonFailedCreatingDatabaseUser,
+				"Failed creating database user. Error: %s", err.Error())
+			return errors.Wrap(err)
+		}
+		logrus.Info("User created successfully")
+	}
+
+	return nil
+}
+
+func extractDBInstanceNames(intents otterizev1alpha3.ClientIntents) []string {
+	dbNames := goset.NewSet[string]()
+	for _, intent := range intents.GetCallsList() {
+		if intent.Type != otterizev1alpha3.IntentTypeDatabase {
+			continue
+		}
+		dbNames.Add(intent.GetTargetServerName())
+	}
+
+	return dbNames.Items()
+}
+
+func findMatchingPGServerConfForDBInstance(
+	databaseInstanceName string,
+	pgServerConfigList otterizev1alpha3.PostgreSQLServerConfigList) (*otterizev1alpha3.PostgreSQLServerConfig, error) {
+
+	for _, conf := range pgServerConfigList.Items {
+		if conf.Name == databaseInstanceName {
+			return &conf, nil
+		}
+	}
```

Comment:
I like `lo.Find` for this (up to you though)

Reply:
I see you applied this change in the intents-operator PR, so make sure you're eventually consistent 🙏